### PR TITLE
feat(common): add unified tokens and unify token func and default tokens

### PR DESCRIPTION
**Pull request by LuPoYi**
- [x] add unified tokens and unify token func
- [x] add default tokens
- [x] test:unit
- [x] yarn changeset

### CODE REVIEW

File: packages/common/src/tokens/utils.ts [hunk 1]
```diff
@@ -1,6 +1,17 @@
...
+  arbitrumTokens,
+  avalancheTokens,
+  baseTokens,
+  gnosisTokens,
+  mainnetTokens,
+  metisTokens,
+  optimismTokens,
+  polygonTokens,
+  zksyncTokens,
+} from './basicTokens';
```
Comment (zodahu) [inline packages/common/src/tokens/utils.ts]: The BNB chain appears to be missing, also in `index.ts`

File: packages/common/src/tokens/basicTokens/index.ts
Comment (zodahu) [inline packages/common/src/tokens/basicTokens/index.ts:1]: Are we expecting to have other folders such as `complexTokens/` and what does `basic` mean? If not, I suggest removing the `basicTokens/` folder and having utils.ts import each JSON directly.

File: packages/common/src/tokens/utils.ts [hunk 2]
```diff
@@ -24,3 +35,25 @@ export function sortByAddress<T extends Token | TokenObject | TokenAmount | Toke
...
+  [ChainId.polygon]: polygonTokens,
+  [ChainId.optimism]: optimismTokens,
+  [ChainId.arbitrum]: arbitrumTokens,
+  [ChainId.metis]: metisTokens,
+  [ChainId.gnosis]: gnosisTokens,
+  [ChainId.zksync]: zksyncTokens,
+  [ChainId.base]: baseTokens,
+};
+
+export function getBasicTokens(chainId: number) {
```
Comment (zodahu) [inline packages/common/src/tokens/utils.ts]: How about renaming it to `getUnifiedTokens` to better align with the `unifyTokens` function?

File: packages/common/src/tokens/utils.test.ts [hunk 3]
```diff
@@ -0,0 +1,30 @@
+import { Token } from './token';
+import { expect } from 'chai';
+import { mainnetTokens } from './basicTokens';
+import { unifyTokens } from './utils';
+
+describe('unifyTokens', () => {
+  it('should replace property if input token is basic token', () => {
+    const chainId = 1;
+    const tokenSet: [number, string, number, string, string, string][] = [
+      [1, '0xA0b86991c6218b36c1d19D4a2e9Eb0cE3606eB48', 6, 'USDC', 'UUUSDC', 'some url'],
```
Comment (zodahu) [inline packages/common/src/tokens/utils.test.ts]:
```suggestion
    const tokenSet = [
      new Token(1, '0xA0b86991c6218b36c1d19D4a2e9Eb0cE3606eB48', 6, 'USDC', 'UUUSDC', 'some url'),
```

File: packages/common/src/tokens/basicTokens/data/arbitrum.json [hunk 4]
```diff
@@ -0,0 +1,202 @@
+{
+  "0xaf88d065e77c8cC2239327C5EDb3A432268e5831": {
```
Comment (zodahu) [inline packages/common/src/tokens/basicTokens/data/arbitrum.json]: Do we expect to use the address as a key and place it in the toTokenMap, assuming the key represents the symbol? Additionally, how should we migrate the token list from each logic (e.g., `protocolink-logics/src/logics/morphoblue/tokens/data/mainnet.json`) to the current format?

File: packages/common/src/tokens/basicTokens/data/zksync.json
Comment (zodahu) [inline packages/common/src/tokens/basicTokens/data/zksync.json:1]: Please help fix zksync token list after the sheet has been corrected

File: packages/common/src/tokens/utils.ts (hunk 2 above)
Comment (LuPoYi) [reply packages/common/src/tokens/utils.ts]: fixed it, thanks for naming sugguestion

File: packages/common/src/tokens/utils.test.ts (hunk 3 above)
Comment (LuPoYi) [reply packages/common/src/tokens/utils.test.ts]: fixed, also put the tokens to outer scope

File: packages/common/src/tokens/utils.ts (hunk 1 above)
Comment (LuPoYi) [reply packages/common/src/tokens/utils.ts]: fixed

File: packages/common/src/tokens/utils.ts (hunk 2 above)
Comment (LuPoYi) [reply packages/common/src/tokens/utils.ts]: changed, this would be better!

File: packages/common/src/tokens/basicTokens/index.ts
Comment (LuPoYi) [reply packages/common/src/tokens/basicTokens/index.ts:1]: fixed

File: packages/common/src/tokens/utils.ts [hunk 5]
```diff
@@ -24,3 +56,86 @@ export function sortByAddress<T extends Token | TokenObject | TokenAmount | Toke
     return address.toLowerCase();
   });
 }
+
+const basicTokenMap: Record<number, Record<string, Token>> = {
```
Comment (zodahu) [inline packages/common/src/tokens/utils.ts]:
```suggestion
const unifiedTokenMap: Record<number, Record<string, Token>> = {
```

also modify the changeset & PR name

File: packages/common/src/tokens/utils.ts [hunk 6]
```diff
@@ -24,3 +56,86 @@ export function sortByAddress<T extends Token | TokenObject | TokenAmount | Toke
...
+  [ChainId.mainnet]: mainnetTokens,
+  [ChainId.avalanche]: avalancheTokens,
+  [ChainId.polygon]: polygonTokens,
+  [ChainId.optimism]: optimismTokens,
+  [ChainId.arbitrum]: arbitrumTokens,
+  [ChainId.metis]: metisTokens,
+  [ChainId.gnosis]: gnosisTokens,
+  [ChainId.zksync]: zksyncTokens,
+  [ChainId.base]: baseTokens,
```
Comment (zodahu) [inline packages/common/src/tokens/utils.ts]: Please add BNB to this list and sort the entries by chain ID to facilitate a clearer code review.

File: packages/common/src/tokens/data/zksync.json [hunk 7]
```diff
@@ -0,0 +1,98 @@
...
+    "chainId": 324,
+    "address": "0x493257fD37EDB34451f62EDf8D2a0C418852bA4C",
+    "decimals": 6,
+    "symbol": "USDT",
+    "name": "Tether USD",
+    "logoUri": "https://cdn.furucombo.app/assets/img/token/USDT.svg"
+  },
+  "0x0000000000000000000000000000000000000000": {
+    "chainId": 324,
+    "address": "0x0000000000000000000000000000000000000000",
```
Comment (zodahu) [inline packages/common/src/tokens/data/zksync.json]:
```suggestion
    "address": "0x000000000000000000000000000000000000800A",
```

Comment (LuPoYi) [reply packages/common/src/tokens/data/zksync.json]: fixed

File: packages/common/src/tokens/utils.ts (hunk 6 above)
Comment (LuPoYi) [reply packages/common/src/tokens/utils.ts]: added

File: packages/common/src/tokens/basicTokens/data/arbitrum.json (hunk 4 above)
Comment (LuPoYi) [reply packages/common/src/tokens/basicTokens/data/arbitrum.json]: as we discussed, I changed `JSON` back to using the token map by symbol, and convert to the token map by address in utils for unifiedTokens

File: packages/common/src/tokens/basicTokens/data/zksync.json
Comment (LuPoYi) [reply packages/common/src/tokens/basicTokens/data/zksync.json:1]: updated

File: packages/common/src/tokens/utils.ts (hunk 5 above)
Comment (LuPoYi) [reply packages/common/src/tokens/utils.ts]: updated

File: packages/common/src/tokens/token.test.ts [hunk 8]
```diff
@@ -232,7 +232,12 @@ describe('Token instance', function () {
 
     testCases.forEach(({ token, expected }, i) => {
       it(`case ${i + 1}`, function () {
-        expect(JSON.stringify(token.unwrapped)).to.eq(JSON.stringify(expected));
+        const unwrapped = token.unwrapped;
+        expect(unwrapped.chainId).to.eq(expected.chainId);
+        expect(unwrapped.address).to.eq(expected.address);
+        expect(unwrapped.decimals).to.eq(expected.decimals);
+        expect(unwrapped.symbol).to.eq(expected.symbol);
+        expect(unwrapped.name).to.eq(expected.name);
```
Comment (zodahu) [inline packages/common/src/tokens/token.test.ts]: Why this `Test unwrapped` can't use `JSON.stringify` but `Test wrapped` can?

File: packages/common/src/web3-toolkit.test.ts [hunk 9]
```diff
@@ -49,7 +48,11 @@ describe('Web3Toolkit', function () {
       it(`${toNetworkId(chainId)}: ${expected.symbol}`, async function () {
         const web3Toolkit = new Web3Toolkit(chainId);
         const token = await web3Toolkit.getToken(tokenAddress);
-        expect(JSON.stringify(token)).to.eq(JSON.stringify(expected));
+        expect(token.chainId).to.eq(expected.chainId);
+        expect(token.address).to.eq(expected.address);
+        expect(token.decimals).to.eq(expected.decimals);
+        expect(token.symbol).to.eq(expected.symbol);
+        expect(token.name).to.eq(expected.name);
```
Comment (zodahu) [inline packages/common/src/web3-toolkit.test.ts]: Can we keep to use `JSON.stringify` but modify the above `expected`?

File: packages/common/src/web3-toolkit.test.ts [hunk 10]
```diff
@@ -93,7 +96,13 @@ describe('Web3Toolkit', function () {
       it(`case ${i + 1}`, async function () {
         const web3Toolkit = new Web3Toolkit(chainId);
         const tokens = await web3Toolkit.getTokens(tokenAddresses);
-        expect(JSON.stringify(tokens)).to.eq(JSON.stringify(expected));
+        for (let i = 0; i < tokens.length; i++) {
+          expect(tokens[i].chainId).to.eq(expected[i].chainId);
+          expect(tokens[i].address).to.eq(expected[i].address);
+          expect(tokens[i].decimals).to.eq(expected[i].decimals);
+          expect(tokens[i].symbol).to.eq(expected[i].symbol);
+          expect(tokens[i].name).to.eq(expected[i].name);
```
Comment (zodahu) [inline packages/common/src/web3-toolkit.test.ts]: Can we keep to use `JSON.stringify` but modify the above `expected`?

File: packages/common/src/tokens/utils.ts (hunk 6 above)
Comment (zodahu) [reply packages/common/src/tokens/utils.ts]: It's still not be sorted yet.

File: packages/common/src/tokens/utils.ts [hunk 11]
```diff
@@ -24,3 +58,131 @@ export function sortByAddress<T extends Token | TokenObject | TokenAmount | Toke
...
+    Array.isArray(Object.values(tokens)[0][0]) &&
+    'address' in Object.values(tokens)[0][0][0]
+  );
+}
+
+export function unifyTokens(chainId: number, tokens: TokenArray): TokenArray;
+export function unifyTokens(chainId: number, tokens: TokenPairArray): TokenPairArray;
+export function unifyTokens(chainId: number, tokens: TokenRecord): TokenRecord;
+export function unifyTokens(chainId: number, tokens: TokenPairRecord): TokenPairRecord;
+export function unifyTokens(chainId: number, tokens: TokenFormat): TokenFormat {
```
Comment (zodahu) [inline packages/common/src/tokens/utils.ts]: If we want to unify simple tokens, atoken and ctokens (including token icons) for those we didn't whitelisted, if those `unifyTokens` functions should use `getDefaultTokens` actually?

File: packages/common/src/tokens/utils.ts (hunk 6 above)
Comment (LuPoYi) [reply packages/common/src/tokens/utils.ts]: sorted

File: packages/common/src/tokens/token.test.ts (hunk 8 above)
Comment (LuPoYi) [reply packages/common/src/tokens/token.test.ts]: fixed with `JSON.stringify`

File: packages/common/src/web3-toolkit.test.ts (hunk 9 above)
Comment (LuPoYi) [reply packages/common/src/web3-toolkit.test.ts]: fixed with `JSON.stringify`

File: packages/common/src/web3-toolkit.test.ts (hunk 10 above)
Comment (LuPoYi) [reply packages/common/src/web3-toolkit.test.ts]: fixed with `JSON.stringify`

File: packages/common/src/tokens/utils.ts (hunk 11 above)
Comment (LuPoYi) [reply packages/common/src/tokens/utils.ts]: Redesign `getUnifiedTokens` - merge `interalTokens` and `externalTokens` (naming is open to suggestion)

**Comment by zodahu** [comment]
LGTM

**Pull request merged**